crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs

```diff
@@ -0,0 +1,527 @@
+                unsafe { condition.run_unsafe((), world) }
+            });
+
+            self.completed_sets.insert(set_idx);
+
+            if !set_conditions_met {
+                // mark all members as completed
+                for sys_idx in schedule.systems_of_sets[set_idx].ones() {
+                    if !self.completed_systems.contains(sys_idx) {
+                        self.skip_system_and_signal_dependents(sys_idx);
+                    }
+                }
+
+                self.completed_sets
+                    .union_with(&schedule.sets_of_sets[set_idx]);
+
+                // restore condition change ticks
+                for condition in set_conditions.iter_mut() {
+                    condition.set_last_change_tick(saved_tick.unwrap());
+                }
+            }
+
+            should_run &= set_conditions_met;
+        }
+
+        if !should_run {
+            return false;
+        }
+
+        let system = schedule.systems[system_index].borrow();
+
+        // evaluate the system's conditions
+        let mut system_conditions = schedule.system_conditions[system_index].borrow_mut();
+        for condition in system_conditions.iter_mut() {
+            condition.set_last_change_tick(system.get_last_change_tick());
+        }
+
+        let should_run = system_conditions.iter_mut().all(|condition| {
+            #[cfg(feature = "trace")]
+            let _condition_span = info_span!("condition", name = &*condition.name()).entered();
+            // SAFETY: access is compatible
+            unsafe { condition.run_unsafe((), world) }
+        });
+
+        if !should_run {
+            self.skip_system_and_signal_dependents(system_index);
+            return false;
+        }
+
+        true
+    }
+
+    fn finish_system_and_signal_dependents(&mut self, system_index: usize) {
+        if !self.system_task_metadata[system_index].is_send {
+            self.local_thread_running = false;
+        }
+
+        if self.system_task_metadata[system_index].is_exclusive {
+            self.exclusive_running = false;
+        }
+
+        self.running_systems.set(system_index, false);
+        self.completed_systems.insert(system_index);
+        self.unapplied_systems.insert(system_index);
+        self.signal_dependents(system_index);
+    }
+
+    fn skip_system_and_signal_dependents(&mut self, system_index: usize) {
+        self.ready_systems.set(system_index, false);
+        self.completed_systems.insert(system_index);
+        self.signal_dependents(system_index);
+    }
+
+    fn signal_dependents(&mut self, system_index: usize) {
+        #[cfg(feature = "trace")]
+        let _span = info_span!("signal_dependents").entered();
+        self.dependents_scratch
+            .extend_from_slice(&self.system_task_metadata[system_index].dependents);
+
+        for &dep_idx in &self.dependents_scratch {
+            let dependent_meta = &mut self.system_task_metadata[dep_idx];
```

Comment:
if we assert that `dep_idx != system_index` we could do this without `dependants_scratch`, though that would require `get_unchecked_mut`.

Alternatively we could maintain a parallel array of `Vec<usize>` for dependencies_remaining instead of embedding it in SystemTaskMetadata, since that's the only part that seems to get mutated over the course of running the executor.

Stretching this further, we could make the `Vec<usize>` a `Vec<AtomicUsize>` and let systems update their dependencies upon completion in the system's task instead forcing this behavior through a channel for synchronization. This does have risks of false sharing, but that might be preferable than strictly serializing dependency management through the parallel executor task. This part isn't blocking and we can try this in a later PR.